File: AuthenticationService.Angular/src/app/services/authentication.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { Observable, of } from 'rxjs';
import { catchError, map } from 'rxjs/operators';

import { Result } from '../models/result/result.model';
import { AppError } from '../models/result/app-error.model';
import { ApiError } from '../models/result/api-error.model';

import { environment } from '../../environments/environment';

import { AuthenticationRequestModel } from '../models/authentication/authentication-request.model';
import { LoginResponseModel } from '../models/authentication/login-response.model';
import { RefreshTokenRequestModel } from '../models/authentication/refresh-token-request.model';
import { ChangePasswordModel } from '../models/authentication/change-password.model';

@Injectable({
  providedIn: 'root',
})
export class AuthenticationService {
  private readonly apiUrl = environment.apiUrl;

  constructor(private readonly http: HttpClient) {}

  LoginAsync(request: AuthenticationRequestModel): Observable<Result<LoginResponseModel>> {
    return this.http.post<LoginResponseModel>(`${this.apiUrl}/Auth/Login`, request).pipe(
      map((data) => Result.Success<LoginResponseModel>(data)),

      catchError((response: HttpErrorResponse) => {
        const statusCode = response.status;

        let apiError: ApiError | null = null;

        if (response.error) {
          if (typeof response.error === 'string') {
            try {
              apiError = JSON.parse(response.error);
            } catch {
              apiError = null;
            }
          } else {
            apiError = response.error as ApiError;
          }
        }

        const error = new AppError(
          apiError?.code ?? 'Unknown',
          apiError?.message ?? 'An unexpected error occured',
          statusCode,
        );

        return of(Result.Failure<LoginResponseModel>(error));
      }),
    );
  }

  RefreshAsync(request: RefreshTokenRequestModel): Observable<Result<LoginResponseModel>> {
    return this.http.post<LoginResponseModel>(`${this.apiUrl}/Auth/Refresh`, request).pipe(
      map((data) => Result.Success<LoginResponseModel>(data)),

      catchError((response: HttpErrorResponse) => {
        const statusCode = response.status;

        let apiError: ApiError | null = null;

        if (response.error) {
          if (typeof response.error === 'string') {
            try {
              apiError = JSON.parse(response.error);
            } catch {
              apiError = null;
            }
          } else {
            apiError = response.error as ApiError;
          }
        }

        const error = new AppError(
          apiError?.code ?? 'Unknown',
          apiError?.message ?? 'An unexpected error occured',
          statusCode,
        );

        return of(Result.Failure<LoginResponseModel>(error));
      }),
    );
  }

  VerifyCredentialsAsync(request: AuthenticationRequestModel): Observable<Result<boolean>> {
    return this.http.post<void>(`${this.apiUrl}/Auth/verify-credentials`, request).pipe(
      map(() => Result.Success<boolean>(true)),

      catchError((response: HttpErrorResponse) => {
        const statusCode = response.status;

        let apiError: ApiError | null = null;

        if (response.error) {
          if (typeof response.error === 'string') {
            try {
              apiError = JSON.parse(response.error);
            } catch {
              apiError = null;
            }
          } else {
            apiError = response.error as ApiError;
          }
        }

        const error = new AppError(
          apiError?.code ?? 'Unknown',
          apiError?.message ?? 'An unexpected error occured',
          statusCode,
        );

        return of(Result.Failure<boolean>(error));
      }),
    );
  }

  ChangePasswordAsync(request: ChangePasswordModel): Observable<Result<boolean>> {
    return this.http.post<void>(`${this.apiUrl}/Auth/change-password`, request).pipe(
      map(() => Result.Success<boolean>(true)),

      catchError((response: HttpErrorResponse) => {
        const statusCode = response.status;

        let apiError: ApiError | null = null;

        if (response.error) {
          if (typeof response.error === 'string') {
            try {
              apiError = JSON.parse(response.error);
            } catch {
              apiError = null;
            }
          } else {
            apiError = response.error as ApiError;
          }
        }

        const error = new AppError(
          apiError?.code ?? 'Unknown',
          apiError?.message ?? 'An unexpected error occured',
          statusCode,
        );

        return of(Result.Failure<boolean>(error));
      }),
    );
  }

  LogoutAsync(request: RefreshTokenRequestModel): Observable<Result<boolean>> {
    return this.http.post<void>(`${this.apiUrl}/Auth/Logout`, request).pipe(
      map(() => Result.Success<boolean>(true)),

      catchError((response: HttpErrorResponse) => {
        const statusCode = response.status;

        let apiError: ApiError | null = null;

        if (response.error) {
          if (typeof response.error === 'string') {
            try {
              apiError = JSON.parse(response.error);
            } catch {
              apiError = null;
            }
          } else {
            apiError = response.error as ApiError;
          }
        }

        const error = new AppError(
          apiError?.code ?? 'Unknown',
          apiError?.message ?? 'An unexpected error occured',
          statusCode,
        );

        return of(Result.Failure<boolean>(error));
      }),
    );
  }
}

File: AuthenticationService.WinForms/Services/AuthenticationService.cs
using AuthenticationService.Dtos.Authentication;
using AuthenticationService.WinForms.Global;
using System.Text.Json;
using System.Text;

namespace AuthenticationService.WinForms.Services
{
    //public static class AuthenticationServiceErrors

    public static class AuthenticationService
    {
        private static readonly HttpClient _Client = new HttpClient()
        {
            BaseAddress = new Uri(Settings.APIBaseURL)
        };

        public static async Task<Result<LoginResponseDto>> LoginAsync(AuthenticationRequestDto request)
        {
            var jsonRequest = JsonSerializer.Serialize(request);

            using var content = new StringContent(jsonRequest, Encoding.UTF8, "application/json");

            HttpResponseMessage response = await _Client.PostAsync("Auth/Login", content);

            if (!response.IsSuccessStatusCode)
            {
                var statusCode = (int)response.StatusCode;

                var jsonResponseError = await response.Content.ReadAsStringAsync();

                var apiError = JsonSerializer.Deserialize<ApiError>(jsonResponseError, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });

                var error = new Error(apiError?.Code ?? "Unknown", apiError?.Message ?? "An unexpected error occured", statusCode);

                return Result<LoginResponseDto>.Failure(error);
            }

            var jsonResponse = await response.Content.ReadAsStringAsync();
            var data = JsonSerializer.Deserialize<LoginResponseDto>(jsonResponse, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });

            return Result<LoginResponseDto>.Success(data);
        }

        public static async Task<Result<LoginResponseDto>> RefreshAsync(RefreshTokenRequestDto request)
        {
            var jsonRequest = JsonSerializer.Serialize(request);

            using var content = new StringContent(jsonRequest, Encoding.UTF8, "application/json");

            HttpResponseMessage response = await _Client.PostAsync("Auth/Refresh", content);

            if (!response.IsSuccessStatusCode)
            {
                var statusCode = (int)response.StatusCode;

                var jsonResponseError = await response.Content.ReadAsStringAsync();

                var apiError = JsonSerializer.Deserialize<ApiError>(jsonResponseError, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });

                var error = new Error(apiError?.Code ?? "Unknown", apiError?.Message ?? "An unexpected error occured", statusCode);

                return Result<LoginResponseDto>.Failure(error);
            }

            var jsonResponse = await response.Content.ReadAsStringAsync();
            var data = JsonSerializer.Deserialize<LoginResponseDto>(jsonResponse, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });

            return Result<LoginResponseDto>.Success(data);
        }

        public static async Task<Result<bool>> VerifyCredentialsAsync(AuthenticationRequestDto request)
        {
            var jsonRequest = JsonSerializer.Serialize(request);

            using var content = new StringContent(jsonRequest, Encoding.UTF8, "application/json");

            HttpResponseMessage response = await _Client.PostAsync("Auth/verify-credentials", content);

            if (!response.IsSuccessStatusCode)
            {
                var statusCode = (int)response.StatusCode;

                var jsonResponseError = await response.Content.ReadAsStringAsync();

                var apiError = JsonSerializer.Deserialize<ApiError>(jsonResponseError, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });

                var error = new Error(apiError?.Code ?? "Unknown", apiError?.Message ?? "An unexpected error occured", statusCode);

                return Result<bool>.Failure(error);
            }

            return Result<bool>.Success(true);
        }

        public static async Task<Result<bool>> ChangePasswordAsync(ChangePasswordDto request)
        {
            var jsonRequest = JsonSerializer.Serialize(request);

            using var content = new StringContent(jsonRequest, Encoding.UTF8, "application/json");

            HttpResponseMessage response = await _Client.PostAsync("Auth/change-password", content);

            if (!response.IsSuccessStatusCode)
            {
                var statusCode = (int)response.StatusCode;

                var jsonResponseError = await response.Content.ReadAsStringAsync();

                var apiError = JsonSerializer.Deserialize<ApiError>(jsonResponseError, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });

                var error = new Error(apiError?.Code ?? "Unknown", apiError?.Message ?? "An unexpected error occured", statusCode);

                return Result<bool>.Failure(error);
            }

            return Result<bool>.Success(true);
        }

        public static async Task<Result<bool>> LogoutAsync(RefreshTokenRequestDto request)
        {
            var jsonRequest = JsonSerializer.Serialize(request);

            using var content = new StringContent(jsonRequest, Encoding.UTF8, "application/json");

            HttpResponseMessage response = await _Client.PostAsync("Auth/logout", content);

            if (!response.IsSuccessStatusCode)
            {
                var statusCode = (int)response.StatusCode;

                var jsonResponseError = await response.Content.ReadAsStringAsync();

                var apiError = JsonSerializer.Deserialize<ApiError>(jsonResponseError, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });

                var error = new Error(apiError?.Code ?? "Unknown", apiError?.Message ?? "An unexpected error occured", statusCode);

                return Result<bool>.Failure(error);
            }

            return Result<bool>.Success(true);
        }
    }
}